In [92]:
import pandas as pd
import numpy as np

# Загрузка и подготовка данных

Импортируем подготовленный датасет.

In [30]:
twitch_df = pd.read_csv('../data/twitch_data_preprocessed.csv')

Убеждаемся, что загрузка прошла успешно.

In [63]:
twitch_df.head()

Watch time(Minutes)  Stream time(minutes)  Peak viewers  Average viewers  \
0           6196161750                215250        222720            27716   
1           6091677300                211845        310998            25610   
2           5644590915                515280        387315            10976   
3           3970318140                517740        300575             7714   
4           3671000070                123660        285644            29602   

   Followers  Views gained  
0    3246298      93036735  
1    5310163      89705964  
2    1767635     102611607  
3    3944850     106546942  
4    8938903      78998587

Создаем датафреймы X (объекты) и y (ответы).

In [32]:
y = twitch_df['Followers gained']
twitch_df = twitch_df.drop(['Followers gained'], axis=1)
X = twitch_df

Разбиваем всю выборку на обучающую и тестовую.

In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 101)

Нормируем вещественные признаки.

In [34]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

# Линейная регрессия

[LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html) - документация по LinearRegression



Импортируем нужные нам модули.

In [35]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

Создаем линейный регрессор, обучаем его с помощью метода fit на тренировочной выборке и предсказываем ответ.

In [36]:
lin_reg  = LinearRegression()
lin_reg.fit(X_train, y_train)
y_pred = lin_reg.predict(X_test)

Оцениваем регрессор с помощью 4 метрик: R2 Score, RMSE, MSE, MAE.

In [37]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5366031433627446
RMSE : 202315.23271199077
Mean Absolute Error: 114457.86433459811
MSE : 40931453387.306984


Результат получился средний, посмотрим, сможем ли мы улучшить его.

# Регрессия дерева решений

[DecisionTreeRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeRegressor.html) - документация по DecisionTreeRegressor

In [38]:
from sklearn.tree import DecisionTreeRegressor

In [39]:
tree_reg = DecisionTreeRegressor()
tree_reg.fit(X_train, y_train)
y_pred = tree_reg.predict(X_test)

In [40]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.42616405731505735
RMSE : 225136.50233238502
Mean Absolute Error: 125712.52
MSE : 50686444682.46


Результат сильно ухудшился, данный регрессор не стоит рассматривать.

# LASSO

[LASSO](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html) - документация по LASSO

In [41]:
from sklearn.linear_model import Lasso

In [42]:
lasso_reg = Lasso()
lasso_reg.fit(X_train, y_train)
y_pred = lasso_reg.predict(X_test)

In [43]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5366032486470997
RMSE : 202315.2097288504
Mean Absolute Error: 114457.65957829492
MSE : 40931444087.62872


Результаты не изменились по сравнению с лучшим полученным ранее регрессором.

# Гребневая регрессия

[Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html) - документация по Ridge

In [44]:
from sklearn.linear_model import Ridge

In [45]:
ridge_reg = Ridge()
ridge_reg.fit(X_train, y_train)
y_pred = ridge_reg.predict(X_test)

In [46]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5367260935538616
RMSE : 202288.3914134878
Mean Absolute Error: 114437.1607948577
MSE : 40920593300.656456


Результаты не изменились по сравнению с лучшим полученным ранее регрессором.

# Elastic Net регрессия

[ElasticNet](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html) - документация по ElasticNet

In [47]:
from sklearn.linear_model import ElasticNet

In [48]:
elastic_net_reg = ElasticNet()
elastic_net_reg.fit(X_train, y_train)
y_pred = elastic_net_reg.predict(X_test)

In [49]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.4665914943444798
RMSE : 217061.10493946433
Mean Absolute Error: 123907.1277858166
MSE : 47115523277.541145


Данные ухудшились, что говорит о том, что данный регрессор не сильно подходит под нашу задачу.

# Случайный лес

In [50]:
from sklearn.ensemble import RandomForestRegressor

In [51]:
random_forest_reg = RandomForestRegressor()
random_forest_reg.fit(X_train, y_train)
y_pred = random_forest_reg.predict(X_test)

In [52]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5541561171613075
RMSE : 198446.50165927145
Mean Absolute Error: 112662.0209
MSE : 39381014020.80323


Качество модели выросло, но не намного, посмотрим, сможем ли мы улучшить результат с помощью подбора гиперпараметров.

# Подбор параметров

Импортируем рандомный поиск с кросс-валидацией.

In [53]:
from sklearn.model_selection import RandomizedSearchCV # импорт нужного модуля


# инициализирование массивов с параметрами
n_estimators = [int(x) for x in np.linspace(start = 100, stop = 1000, num = 10)]
max_features = ['log2', 'sqrt']
max_depth = [int(x) for x in np.linspace(start = 1, stop = 15, num = 15)]
min_samples_split = [int(x) for x in np.linspace(start = 2, stop = 50, num = 10)]
min_samples_leaf = [int(x) for x in np.linspace(start = 2, stop = 50, num = 10)]
bootstrap = [True, False]


# словарь с параметрами
param_dist = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}


# запускаем рандомный поиск гиперпараметров
rs = RandomizedSearchCV(random_forest_reg, 
                        param_dist, 
                        n_iter = 100, 
                        cv = 3, 
                        verbose = 1, 
                        n_jobs=-1, 
                        random_state=0)
rs.fit(X_train, y_train)
rs.best_params_

Fitting 3 folds for each of 100 candidates, totalling 300 fits


{'n_estimators': 500,
 'min_samples_split': 39,
 'min_samples_leaf': 18,
 'max_features': 'sqrt',
 'max_depth': 11,
 'bootstrap': False}

Предполагаемые гиперпараметры найдены, проверим, улучшат ли они качество модели.

In [54]:
random_forest_reg = RandomForestRegressor(n_estimators=500, min_samples_split=39, min_samples_leaf=18,max_features='sqrt', max_depth=11, bootstrap=False)
random_forest_reg.fit(X_train, y_train)
y_pred = random_forest_reg.predict(X_test)

In [55]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.4990836360416526
RMSE : 210346.19039682223
Mean Absolute Error: 119534.5953575917
MSE : 44245519814.45619


Как видно по метрике R2 Score, качество модели заметно ухудшилось, что говорит о том, что данные гиперпараметры не подходят под данную задачу. Попробуем поиск с помощью сетки.

In [58]:
from sklearn.model_selection import GridSearchCV

random_forest_reg = RandomForestRegressor()

param_dist = {'n_estimators': range(10, 51, 10),
               'max_depth': range(1, 13, 4),
               'min_samples_split': range(2, 10, 4),
               'min_samples_leaf': range(1, 8, 4)}

clf = GridSearchCV(random_forest_reg, param_dist, cv=5)

clf.fit(X_train, y_train)

clf.best_params_

{'max_depth': 5,
 'min_samples_leaf': 5,
 'min_samples_split': 2,
 'n_estimators': 10}

In [59]:
random_forest_reg = RandomForestRegressor(n_estimators=10, min_samples_split=2, min_samples_leaf=5, max_depth=5)
random_forest_reg.fit(X_train, y_train)
y_pred = random_forest_reg.predict(X_test)

In [60]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5390612743893326
RMSE : 201777.91937575507
Mean Absolute Error: 114082.33748091575
MSE : 40714328747.60871


In [61]:
random_forest_reg = RandomForestRegressor(n_estimators=50, min_samples_split=2, min_samples_leaf=5, max_depth=9)
random_forest_reg.fit(X_train, y_train)
y_pred = random_forest_reg.predict(X_test)

In [62]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5463693761144289
RMSE : 200171.9518013022
Mean Absolute Error: 113251.85901646194
MSE : 40068810287.942856


Как видно по метрикам, данные гиперпараметры не только не улучшили модель, но и слегка ухудшили ее. Попробуем поискать гиперпараметры для других моделей.

In [79]:
# инициализирование массивов с параметрами
max_iter = [int(x) for x in np.linspace(start = 1, stop = 1000, num = 10)]
alpha = [int(x) for x in np.linspace(start = 1, stop = 1000, num = 10)]


# словарь с параметрами
param_dist = {'alpha': alpha,
               'max_iter': max_iter}


# запускаем рандомный поиск гиперпараметров
rs = RandomizedSearchCV(lasso_reg, 
                        param_dist, 
                        n_iter = 100, 
                        cv = 3, 
                        verbose = 1, 
                        n_jobs=-1, 
                        random_state=0)
rs.fit(X_train, y_train)
rs.best_params_

Fitting 3 folds for each of 100 candidates, totalling 300 fits


{'max_iter': 112, 'alpha': 1000}

In [80]:
lasso_reg = Lasso(max_iter=112, alpha=1000)
lasso_reg.fit(X_train, y_train)
y_pred = lasso_reg.predict(X_test)

In [81]:
print(f'R2 score: {r2_score(y_test, y_pred)}')
print(f'RMSE : {np.sqrt(mean_squared_error(y_test, y_pred))}')
print(f'Mean Absolute Error: {mean_absolute_error(y_test, y_pred)}')
print(f'MSE : {mean_squared_error(y_test, y_pred)}')

R2 score: 0.5384073883127406
RMSE : 201920.98938218964
Mean Absolute Error: 114055.07627742253
MSE : 40772085953.08234


Видим, что модель практически не изменилась, попробуем подобрать параметры с помощью сетки.

In [82]:
clf = GridSearchCV(lasso_reg, param_dist, cv=5)

clf.fit(X_train, y_train)

clf.best_params_

x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 24724561422080.48, tolerance: 8749906976.141245
  model = cd_fast.enet_coordinate_descent(
x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 24079489876416.24, tolerance: 8482990220.98435
  model = cd_fast.enet_coordinate_descent(
x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 14897678026194.205, tolerance: 5650638912.045841
  model = cd_fast.enet_coordinate_descent(
x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not con

x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 23663573166202.86, tolerance: 8482990220.98435
  model = cd_fast.enet_coordinate_descent(
x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 14451858415522.082, tolerance: 5650638912.045841
  model = cd_fast.enet_coordinate_descent(
x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 18663311560312.7, tolerance: 7480341758.200807
  model = cd_fast.enet_coordinate_descent(
x:\programs\python\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not conv

{'alpha': 1000, 'max_iter': 112}

Сетка выдала те же самые результаты.

# Итоги

Какие бы гиперпараметры не подбирались, лучше стандартного случайного леса ничего не оказалось. Результат, оцениваемый метрикой R2 в примерно 0.55, оказался лучшим в данной ситуации.